keps/sig-multicluster/4322-cluster-inventory/README.md

```diff
@@ -0,0 +1,993 @@
+-->
+
+* Provide a standard reference implementation.
+* Define specific implementation details beyond general API behavior.
+* Offering functionalities related to multi-cluster orchestration.
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+the API proposed by this KEP aims to
+- Provide a reliable, consistent, and automated approach for any
+  multi-cluster application (framework, toolset) to discover available
+  clusters and take actions accordingly, in a way similar to service
+  discovery works in a microservice architecture. Through the inventory,
+  the application can query for a list of clusters to access, or watch
+  for an ever-flowing stream of cluster lifecycle events which the
+  application can act upon timely, such as auto-scaling, upgrades,
+  failures, and connectivity issues.
+- Provide a simple, clear interface for human operators to understand
+  clusters under management, and troubleshoot them in a streamlined
+  manner.
+
+### Terminology
+
+- **Cluster Manager**: An entity that creates the cluster inventory API 
+  resources per member cluster, and keeps their status up-to-date. Each
+  cluster manager SHOULD be identified with a unique name. Each cluster
+  inventory resource SHOULD be managed by only one cluster manager. Examples
+  of cluster manager are projects like OCM, Karmada, Clusternet or Azure
+  fleet manager.
+
+- **Cluster Inventory Consumer**: Users or third party tools that use
+  cluster inventory API for the purpose of workload distribution, operation
+  management etc. The consumer SHOULD not create this API resource.
+
+- **Member Cluster**: A kubernetes cluster that is managed by the cluster
+manager. A cluster manager SHOULD have sufficient permission to access
+the member cluster to fetch the information so it can update the status
+of the cluster inventory API resource.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1: Multicluster Workload Distribution
+
+In this scenario, the unified API acts as a foundation for multicluster
+scheduling across various clusters. This means that the API will provide
+a standardized way to distribute workloads across multiple clusters. For
+instance, workload distribution tools like GitOps or
+[Work API](https://github.com/kubernetes-sigs/work-api) can leverage this
+API to make informed decisions about which cluster is best suited to handle
+a particular workload.
```

Comment:
```suggestion
In this scenario, a standard API acts as a foundation for multicluster
scheduling decisions across various clusters. This means that the 
API will provide a single point of contact to retrieve information that can be
used to make informed scheduling decisions for workloads across multiple
clusters. For instance, workload distribution tools like GitOps or
[Work API](https://github.com/kubernetes-sigs/work-api) can leverage this
API to make informed decisions about which cluster is best suited to handle
a particular workload.
```

Reply:
As  it reads today sounds like the API can be used for cluster management `a standardized way to distribute workloads across multiple clusters.` that's not true. the `ClusterInventory` provides information, then Schedulers make the decision.

Reply:
AFAIK Kubernetes scheduler uses API fields like taints/tolerations.  I think that a multicluster scheduler will need to leverage API fields like this to make decisions about placement of the workload.

My main thoughts are that if this is a user story as part of this KEP, I think this should be targeted in alpha.  So maybe there should be more thoughts on the API for mutlicluster workload distribution.

Reply:
@kannon92 We discussed the taint before and finally decide to put it in the beta stage. The reason is taint and tolerations are pairs and it is not easy to shape taint if toleration is missing. Toleration is most likely in another API relating to multicluster scheduling which would be another huge topic. Hence we hope to build this API as common & simple as possible at the first stage, and continuously iterate it. WDYT?

Reply:
@ArangoGutierrez this story is updated

Reply:
As discussed on the `sig-multicluster` call, a strategy of ` continuously iterate it. WDYT?` is not appropriate for a new API, if the first version is of no use for the community, no one will participate in the conversation moving forward, and the iterations won't be successful.  We know there's a lot of need in the community for an Unified API for MultiCluster inventory, but v1alpha1 should serve a broad set of users from the beginning.

Reply:
> @kannon92 We discussed the taint before and finally decide to put it in the beta stage. The reason is taint and tolerations are pairs and it is not easy to shape taint if toleration is missing. Toleration is most likely in another API relating to multicluster scheduling which would be another huge topic. Hence we hope to build this API as common & simple as possible at the first stage, and continuously iterate it. WDYT?

This sounds good to me. From what I understand, this KEP will be iterated outside of core k8s.  I think I'd like to see high level goals for this project on this KEP as I'm sure the API will be iterated on in the subproject.